test/test_foreach.py

```diff
@@ -149,13 +163,106 @@ def _test_bin_op_list_alpha(self, device, dtype, foreach_op, foreach_op_, torch_
+                if (dtype is torch.float16 or dtype is torch.bfloat16) and TEST_WITH_ROCM:
+                    self.assertEqual(res, expected, atol=1.e-3, rtol=self.dtype_precisions[dtype][0])
 
-    @dtypes(*[torch.float, torch.double, torch.complex64, torch.complex128])
-    def test_exp(self, device, dtype):
-        self._test_unary_op(device, dtype, torch._foreach_exp, torch._foreach_exp_, torch.exp)
+                    fe_op_(tensors1)
+                    self.assertEqual(res, tensors1)
+                else:
+                    self.assertEqual(res, expected)
+
+                    fe_op_(tensors1)
+                    self.assertEqual(res, tensors1)
+
+    # Separate test for abs due to a lot of special cases
+    # Absolute value of a complex number a + bj is defined as sqrt(a^2 + b^2), i.e. a floating point
+    @dtypes(*(torch.testing.floating_and_complex_types_and(torch.bfloat16, torch.half)))
+    def test_abs(self, device, dtype):
+        for N in N_values:
+            tensors1 = self._get_test_data(device, dtype, N)
+            # Mimics cuda kernel dtype flow.  With fp16/bf16 input, runs in fp32 and casts output back to fp16/bf16.
+            control_dtype = torch.float32 if (self.device_type == 'cuda' and
+                                              (dtype is torch.float16 or dtype is torch.bfloat16)) else dtype
+
+            expected = [torch.abs(tensors1[i].to(dtype=control_dtype)).to(dtype=dtype) for i in range(N)]
+            res = torch._foreach_abs(tensors1)
+            if (dtype is torch.float16 or dtype is torch.bfloat16) and TEST_WITH_ROCM:
+                self.assertEqual(res, expected, atol=1.e-3, rtol=self.dtype_precisions[dtype][0])
+
+                torch._foreach_abs_(tensors1)
+                self.assertEqual(res, tensors1)
+            else:
+                if self.device_type == 'cpu':
+                    if dtype == torch.complex64:
+                        expected = [torch.abs(tensors1[i]) for i in range(N)]
+                        self.assertEqual(res, expected)
+
+                        with self.assertRaisesRegex(RuntimeError, r"In-place abs is not supported for complex tensors."):
+                            torch._foreach_abs_(tensors1)
+                        break
+                    elif dtype == torch.complex128:
+                        expected = [torch.abs(tensors1[i]) for i in range(N)]
+                        self.assertEqual(res, expected)
+
+                        with self.assertRaisesRegex(RuntimeError, r"In-place abs is not supported for complex tensors."):
+                            torch._foreach_abs_(tensors1)
+                        break
+                    else:
+                        self.assertEqual(res, expected)
+                else:
+                    if dtype == torch.complex64:
+                        expected = [torch.abs(tensors1[i]).to(torch.complex64) for i in range(N)]
+                        self.assertEqual(res, expected)
```

Comment:
res computed as `res = torch._foreach_abs(tensors1)` has real datatype for the tensors in the list, whereas the above line makes each tensor in the `expected` list a complex tensor. 
That test is failing on ROCm because of this type difference. Anything I am missing or is behaving differently than it should on ROCm?

Reply:
in case of complex input, the result will be of real type. Currently, foreach APIs dont support different return type. As a fix we are pushing complex to a slow path.